curp/src/sync_manager.rs

```diff
@@ -144,137 +137,222 @@ impl<C: Command + 'static> SyncManager<C> {
-                        connect
-                            .commit(CommitRequest::new(
-                                term,
-                                index.numeric_cast(),
-                                cmds_cloned.as_ref(),
-                            )?)
-                            .await
-                    },
-                    |r| match r.into_inner().commit_response {
-                        Some(CommitResponse::Committed(_)) => 1,
-                        Some(CommitResponse::WrongTerm(_) | CommitResponse::PrevNotReady(_)) => 0,
-                        None => unreachable!("Should contain sync response"),
-                    },
-                    || true,
-                    max_fail,
-                )
-                .await
-            });
+        // send append_entries request and receive response
+        debug!("append_entries sent to {}", connect.addr);
+        let resp = connect
+            .append_entries(req, Self::APPEND_ENTRIES_TIMEOUT)
+            .await;
+
+        #[allow(clippy::unwrap_used)]
+        // indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+        match resp {
+            Err(e) => warn!("Heartbeat error: {}", e),
+            Ok(resp) => {
+                // TODO: update term
+                let resp = resp.into_inner();
+                let mut state = state.write();
+                if resp.success {
+                    // update match_index and next_index
+                    *state.match_index.get_mut(&connect.addr).unwrap() = last_sent_index;
+                    *state.next_index.get_mut(&connect.addr).unwrap() = last_sent_index + 1;
+
+                    // update commit_index if needed
+                    let majority_cnt = (state.others.len() + 1) / 2 + 1;
+                    for i in ((state.commit_index + 1)..=state.last_log_index()).rev() {
+                        // should not commit logs from previous term
+                        if state.log[i].term() != state.term {
+                            break;
+                        }
+
+                        // calculate the number of followers that has matched log[i]
+                        let match_count = state
+                            .others
+                            .iter()
+                            .filter(|addr| state.match_index[*addr] >= i)
+                            .count();
+
+                        // If the majority of servers has replicated the log, commit
+                        // + 1 means including the leader itself
+                        if match_count + 1 >= majority_cnt {
+                            debug!("commit_index updated to {i}");
+                            state.commit_index = i;
+                            commit_notify.notify_one();
+                            ae_notify.notify_one();
+                            break;
+                        }
+                    }
```

Comment:
The algorithm is not efficient. We don't have to call iter().filter() for possible index. As we're holding the state write lock, the can calculate the index by scanning the match_index once.